// moralis/lib/react-native/MoralisWeb3.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.EthereumEvents = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _construct2 = _interopRequireDefault(require("@babel/runtime/helpers/construct"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _web = _interopRequireDefault(require("web3"));

var _ParseObject = _interopRequireDefault(require("./ParseObject"));

var _ParseQuery = _interopRequireDefault(require("./ParseQuery"));

var _ParseUser = _interopRequireDefault(require("./ParseUser"));

var _ParseACL = _interopRequireDefault(require("./ParseACL"));

var _MoralisErd = _interopRequireDefault(require("./MoralisErd"));

var _MoralisDot = _interopRequireDefault(require("./MoralisDot"));

var _MoralisSol = _interopRequireDefault(require("./MoralisSol"));

var _MoralisWalletConnectProvider = _interopRequireDefault(require("./MoralisWalletConnectProvider"));

var _MoralisCustomProvider = _interopRequireDefault(require("./MoralisCustomProvider"));

var _MoralisInjectedProvider = _interopRequireDefault(require("./MoralisInjectedProvider"));

var _TransferUtils = _interopRequireDefault(require("./TransferUtils"));

var _Cloud = require("./Cloud");

var _detectProvider = _interopRequireDefault(require("@metamask/detect-provider"));

var _createSigningData = _interopRequireDefault(require("./createSigningData"));

function _createForOfIteratorHelperLoose(o, allowArrayLike) {
  var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"];
  if (it) return (it = it.call(o)).next.bind(it);

  if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
    if (it) o = it;
    var i = 0;
    return function () {
      if (i >= o.length) return {
        done: true
      };
      return {
        done: false,
        value: o[i++]
      };
    };
  }

  throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) {
    arr2[i] = arr[i];
  }

  return arr2;
}

var EventEmitter = require('events');

var EthereumEvents = {
  CONNECT: 'connect',
  DISCONNECT: 'disconnect',
  ACCOUNTS_CHANGED: 'accountsChanged',
  CHAIN_CHANGED: 'chainChanged'
};
exports.EthereumEvents = EthereumEvents;
var WARNING = 'Non ethereum enabled browser';
var ERROR_WEB3_MISSING = 'Missing web3 instance, make sure to call Moralis.enableWeb3() or Moralis.authenticate()';

var MoralisWeb3 = function () {
  function MoralisWeb3() {
    (0, _classCallCheck2.default)(this, MoralisWeb3);
    var MWeb3 = typeof _web.default === 'function' ? _web.default : window.Web3;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return (0, _construct2.default)(MWeb3, args);
  }

  (0, _createClass2.default)(MoralisWeb3, null, [{
    key: "isWeb3Enabled",
    value: function () {
      return this.ensureWeb3IsInstalled();
    }
  }, {
    key: "setEnableWeb3",
    value: function (fn) {
      this.customEnableWeb3 = fn;
    }
  }, {
    key: "enableWeb3",
    value: function (options) {
      var web3, Web3Provider, web3Provider;
      return _regenerator.default.async(function (_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (!this.customEnableWeb3) {
                _context.next = 6;
                break;
              }

              _context.next = 3;
              return _regenerator.default.awrap(this.customEnableWeb3(options));

            case 3:
              web3 = _context.sent;
              _context.next = 13;
              break;

            case 6:
              if (this.speedyNodeApiKey) {
                options.speedyNodeApiKey = this.speedyNodeApiKey;
                options.provider = 'custom';
              }

              Web3Provider = MoralisWeb3.getWeb3Provider(options);
              web3Provider = new Web3Provider();
              this.activeWeb3Provider = web3Provider;
              _context.next = 12;
              return _regenerator.default.awrap(web3Provider.activate(options));

            case 12:
              web3 = _context.sent;

            case 13:
              this.web3 = web3;
              return _context.abrupt("return", web3);

            case 15:
            case "end":
              return _context.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "enable",
    value: function (options) {
      return _regenerator.default.async(function (_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              console.warn('Moralis.enable() is deprecated and will be removed, use Moralis.enableWeb3() instead.');
              return _context2.abrupt("return", this.enableWeb3(options));

            case 2:
            case "end":
              return _context2.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "isDotAuth",
    value: function (options) {
      switch (options == null ? void 0 : options.type) {
        case 'dot':
        case 'polkadot':
        case 'kusama':
          return true;

        default:
          return false;
      }
    }
  }, {
    key: "isElrondAuth",
    value: function (options) {
      switch (options == null ? void 0 : options.type) {
        case 'erd':
        case 'elrond':
          return true;

        default:
          return false;
      }
    }
  }, {
    key: "isSolAuth",
    value: function (options) {
      switch (options == null ? void 0 : options.type) {
        case 'sol':
          return true;

        default:
          return false;
      }
    }
  }, {
    key: "getWeb3Provider",
    value: function (options) {
      switch (options == null ? void 0 : options.provider) {
        case 'walletconnect':
        case 'walletConnect':
        case 'wc':
          return _MoralisWalletConnectProvider.default;

        case 'custom':
          return _MoralisCustomProvider.default;

        default:
          return _MoralisInjectedProvider.default;
      }
    }
  }, {
    key: "cleanup",
    value: function () {
      return _regenerator.default.async(function (_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              if (!this.activeWeb3Provider) {
                _context3.next = 3;
                break;
              }

              _context3.next = 3;
              return _regenerator.default.awrap(this.activeWeb3Provider.deactivate());

            case 3:
              _MoralisWalletConnectProvider.default.cleanupStaleData();

            case 4:
            case "end":
              return _context3.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "authenticate",
    value: function (options) {
      var isLoggedIn, web3, message, data, accounts, accountsLower, _accountsLower, ethAddress, signature, authData, user;

      return _regenerator.default.async(function (_context4) {
        while (1) {
          switch (_context4.prev = _context4.next) {
            case 0:
              _context4.next = 2;
              return _regenerator.default.awrap(_ParseUser.default.currentAsync());

            case 2:
              isLoggedIn = _context4.sent;

              if (!isLoggedIn) {
                _context4.next = 6;
                break;
              }

              _context4.next = 6;
              return _regenerator.default.awrap(_ParseUser.default.logOut());

            case 6:
              _context4.next = 8;
              return _regenerator.default.awrap(MoralisWeb3.cleanup());

            case 8:
              if (!MoralisWeb3.isDotAuth(options)) {
                _context4.next = 10;
                break;
              }

              return _context4.abrupt("return", _MoralisDot.default.authenticate(options));

            case 10:
              if (!MoralisWeb3.isElrondAuth(options)) {
                _context4.next = 12;
                break;
              }

              return _context4.abrupt("return", _MoralisErd.default.authenticate(options));

            case 12:
              if (!MoralisWeb3.isSolAuth(options)) {
                _context4.next = 14;
                break;
              }

              return _context4.abrupt("return", _MoralisSol.default.authenticate(options));

            case 14:
              _context4.next = 16;
              return _regenerator.default.awrap(this.enableWeb3(options));

            case 16:
              web3 = _context4.sent;
              message = (options == null ? void 0 : options.signingMessage) || MoralisWeb3.getSigningData();
              _context4.next = 20;
              return _regenerator.default.awrap((0, _createSigningData.default)(message));

            case 20:
              data = _context4.sent;
              _context4.next = 23;
              return _regenerator.default.awrap(web3.eth.getAccounts());

            case 23:
              accounts = _context4.sent;
              accountsLower = accounts.map(function (v) {
                return v.toLowerCase();
              });
              _accountsLower = (0, _slicedToArray2.default)(accountsLower, 1), ethAddress = _accountsLower[0];

              if (ethAddress) {
                _context4.next = 28;
                break;
              }

              throw new Error('Address not found');

            case 28:
              _context4.next = 30;
              return _regenerator.default.awrap(web3.eth.personal.sign(data, ethAddress, ''));

            case 30:
              signature = _context4.sent;

              if (signature) {
                _context4.next = 33;
                break;
              }

              throw new Error('Data not signed');

            case 33:
              authData = {
                id: ethAddress,
                signature: signature,
                data: data
              };
              _context4.next = 36;
              return _regenerator.default.awrap(_ParseUser.default.logInWith('moralisEth', {
                authData: authData
              }));

            case 36:
              user = _context4.sent;

              if (user) {
                _context4.next = 39;
                break;
              }

              throw new Error('Could not get user');

            case 39:
              _context4.next = 41;
              return _regenerator.default.awrap(user.setACL(new _ParseACL.default(user)));

            case 41:
              user.addAllUnique('accounts', accountsLower);
              user.set('ethAddress', ethAddress);
              _context4.next = 45;
              return _regenerator.default.awrap(user.save(null, options));

            case 45:
              return _context4.abrupt("return", user);

            case 46:
            case "end":
              return _context4.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "link",
    value: function (account, options) {
      var web3, data, user, ethAddress, EthAddress, query, ethAddressRecord, signature, authData;
      return _regenerator.default.async(function (_context5) {
        while (1) {
          switch (_context5.prev = _context5.next) {
            case 0:
              _context5.next = 2;
              return _regenerator.default.awrap(MoralisWeb3.enableWeb3(options));

            case 2:
              web3 = _context5.sent;
              data = (options == null ? void 0 : options.signingMessage) || MoralisWeb3.getSigningData();
              _context5.next = 6;
              return _regenerator.default.awrap(_ParseUser.default.currentAsync());

            case 6:
              user = _context5.sent;
              ethAddress = account.toLowerCase();
              EthAddress = _ParseObject.default.extend('_EthAddress');
              query = new _ParseQuery.default(EthAddress);
              _context5.next = 12;
              return _regenerator.default.awrap(query.get(ethAddress).catch(function () {
                return null;
              }));

            case 12:
              ethAddressRecord = _context5.sent;

              if (ethAddressRecord) {
                _context5.next = 20;
                break;
              }

              _context5.next = 16;
              return _regenerator.default.awrap(web3.eth.personal.sign(data, account, ''));

            case 16:
              signature = _context5.sent;
              authData = {
                id: ethAddress,
                signature: signature,
                data: data
              };
              _context5.next = 20;
              return _regenerator.default.awrap(user.linkWith('moralisEth', {
                authData: authData
              }));

            case 20:
              user.addAllUnique('accounts', [ethAddress]);
              user.set('ethAddress', ethAddress);
              _context5.next = 24;
              return _regenerator.default.awrap(user.save(null, options));

            case 24:
              return _context5.abrupt("return", user);

            case 25:
            case "end":
              return _context5.stop();
          }
        }
      }, null, null, null, Promise);
    }
  }, {
    key: "unlink",
    value: function (account) {
      var _user$get;

      var accountsLower, EthAddress, query, ethAddressRecord, user, accounts, nextAccounts;
      return _regenerator.default.async(function (_context6) {
        while (1) {
          switch (_context6.prev = _context6.next) {
            case 0:
              accountsLower = account.toLowerCase();
              EthAddress = _ParseObject.default.extend('_EthAddress');
              query = new _ParseQuery.default(EthAddress);
              _context6.next = 5;
              return _regenerator.default.awrap(query.get(accountsLower));

            case 5:
              ethAddressRecord = _context6.sent;
              _context6.next = 8;
              return _regenerator.default.awrap(ethAddressRecord.destroy());

            case 8:
              _context6.next = 10;
              return _regenerator.default.awrap(_ParseUser.default.currentAsync());

            case 10:
              user = _context6.sent;
              accounts = (_user$get = user.get('accounts')) != null ? _user$get : [];
              nextAccounts = accounts.filter(function (v) {
                return v !== accountsLower;
              });
              user.set('accounts', nextAccounts);
              user.set('ethAddress', nextAccounts[0]);
              _context6.next = 17;
              return _regenerator.default.awrap(user._unlinkFrom('moralisEth'));

            case 17:
              _context6.next = 19;
              return _regenerator.default.awrap(user.save());

            case 19:
              return _context6.abrupt("return", user);

            case 20:
            case "end":
              return _context6.stop();
          }
        }
      }, null, null, null, Promise);
    }
  }, {
    key: "initPlugins",
    value: function (installedPlugins) {
      var _this = this;

      var specs, allPlugins;
      return _regenerator.default.async(function (_context8) {
        while (1) {
          switch (_context8.prev = _context8.next) {
            case 0:
              _context8.t0 = installedPlugins;

              if (_context8.t0) {
                _context8.next = 5;
                break;
              }

              _context8.next = 4;
              return _regenerator.default.awrap((0, _Cloud.run)('getPluginSpecs'));

            case 4:
              _context8.t0 = _context8.sent;

            case 5:
              specs = _context8.t0;
              if (!this.Plugins) this.Plugins = {};

              if (specs) {
                _context8.next = 9;
                break;
              }

              return _context8.abrupt("return");

            case 9:
              allPlugins = this.Plugins;
              specs.forEach(function (plugin) {
                allPlugins[plugin.name] = {};
                plugin.functions.forEach(function (f) {
                  allPlugins[plugin.name][f] = function (params, options) {
                    var response, error, triggerReturn;
                    return _regenerator.default.async(function (_context7) {
                      while (1) {
                        switch (_context7.prev = _context7.next) {
                          case 0:
                            if (!options) options = {};
                            _context7.next = 3;
                            return _regenerator.default.awrap((0, _Cloud.run)(plugin.name + "_" + f, params));

                          case 3:
                            response = _context7.sent;

                            if (response.data.success) {
                              _context7.next = 7;
                              break;
                            }

                            error = JSON.stringify(response.data.data, null, 2);
                            throw new Error("Something went wrong\n" + error);

                          case 7:
                            if (!(options.disableTriggers !== true)) {
                              _context7.next = 13;
                              break;
                            }

                            _context7.next = 10;
                            return _regenerator.default.awrap(_this.handleTriggers(response.data.result.triggers, response.data.result.data));

                          case 10:
                            triggerReturn = _context7.sent;

                            if (!triggerReturn) {
                              _context7.next = 13;
                              break;
                            }

                            return _context7.abrupt("return", triggerReturn);

                          case 13:
                            return _context7.abrupt("return", response.data.result);

                          case 14:
                          case "end":
                            return _context7.stop();
                        }
                      }
                    }, null, null, null, Promise);
                  };
                });
              });
              this.Plugins = allPlugins;

            case 12:
            case "end":
              return _context8.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "handleTriggers",
    value: function (triggersArray, payload) {
      var _this2 = this;

      var response, _loop, i, _ret;

      return _regenerator.default.async(function (_context12) {
        while (1) {
          switch (_context12.prev = _context12.next) {
            case 0:
              if (triggersArray) {
                _context12.next = 2;
                break;
              }

              return _context12.abrupt("return");

            case 2:
              _loop = function (i) {
                var _triggersArray$i, _triggersArray$i2, _triggersArray$i3, _triggersArray$i4, _triggersArray$i5, _triggersArray$i6, _triggersArray$i7, _triggersArray$i8, _triggersArray$i9, _triggersArray$i10, _triggersArray$i11, _triggersArray$i12, _triggersArray$i13, _triggersArray$i14, _triggersArray$i15, _triggersArray$i16, _triggersArray$i17, _triggersArray$i18, _triggersArray$i19, _triggersArray$i21, _triggersArray$i23, _triggersArray$i24, _triggersArray$i25;

                return _regenerator.default.async(function (_context11) {
                  while (1) {
                    switch (_context11.prev = _context11.next) {
                      case 0:
                        _context11.t0 = triggersArray[i].name;
                        _context11.next = _context11.t0 === 'openUrl' ? 3 : _context11.t0 === 'web3Transaction' ? 5 : _context11.t0 === 'web3Sign' ? 18 : _context11.t0 === 'callPluginEndpoint' ? 35 : _context11.t0 === 'web3SignV4' ? 57 : 72;
                        break;

                      case 3:
                        if (triggersArray[i].newTab) window.open(triggersArray[i].url);else window.open(triggersArray[i].url, '_self');
                        return _context11.abrupt("break", 73);

                      case 5:
                        if (_this2.ensureWeb3IsInstalled()) {
                          _context11.next = 7;
                          break;
                        }

                        throw new Error(ERROR_WEB3_MISSING);

                      case 7:
                        if (!(((_triggersArray$i = triggersArray[i]) == null ? void 0 : _triggersArray$i.shouldAwait) === true)) {
                          _context11.next = 11;
                          break;
                        }

                        _context11.next = 10;
                        return _regenerator.default.awrap(_this2.web3.eth.sendTransaction((_triggersArray$i2 = triggersArray[i]) == null ? void 0 : _triggersArray$i2.data));

                      case 10:
                        response = _context11.sent;

                      case 11:
                        if (((_triggersArray$i3 = triggersArray[i]) == null ? void 0 : _triggersArray$i3.shouldAwait) === false) response = _this2.web3.eth.sendTransaction((_triggersArray$i4 = triggersArray[i]) == null ? void 0 : _triggersArray$i4.data);
                        if (((_triggersArray$i5 = triggersArray[i]) == null ? void 0 : _triggersArray$i5.saveResponse) === true) _this2.memoryCard.save(response);

                        if (!(((_triggersArray$i6 = triggersArray[i]) == null ? void 0 : _triggersArray$i6.shouldReturnPayload) === true)) {
                          _context11.next = 15;
                          break;
                        }

                        return _context11.abrupt("return", {
                          v: {
                            payload: payload,
                            response: response
                          }
                        });

                      case 15:
                        if (!(((_triggersArray$i7 = triggersArray[i]) == null ? void 0 : _triggersArray$i7.shouldReturnResponse) === true)) {
                          _context11.next = 17;
                          break;
                        }

                        return _context11.abrupt("return", {
                          v: response
                        });

                      case 17:
                        return _context11.abrupt("break", 73);

                      case 18:
                        if (_this2.ensureWeb3IsInstalled()) {
                          _context11.next = 20;
                          break;
                        }

                        throw new Error(ERROR_WEB3_MISSING);

                      case 20:
                        if (triggersArray[i].message) {
                          _context11.next = 22;
                          break;
                        }

                        throw new Error('web3Sign trigger does not have a message to sign');

                      case 22:
                        if (!(!triggersArray[i].signer || !_this2.web3.utils.isAddress(triggersArray[i].signer))) {
                          _context11.next = 24;
                          break;
                        }

                        throw new Error('web3Sign trigger signer address missing or invalid');

                      case 24:
                        if (!(((_triggersArray$i8 = triggersArray[i]) == null ? void 0 : _triggersArray$i8.shouldAwait) === true)) {
                          _context11.next = 28;
                          break;
                        }

                        _context11.next = 27;
                        return _regenerator.default.awrap(_this2.web3.eth.personal.sign(triggersArray[i].message, triggersArray[i].signer));

                      case 27:
                        response = _context11.sent;

                      case 28:
                        if (((_triggersArray$i9 = triggersArray[i]) == null ? void 0 : _triggersArray$i9.shouldAwait) === false) response = _this2.web3.eth.personal.sign(triggersArray[i].message, triggersArray[i].signer);
                        if (((_triggersArray$i10 = triggersArray[i]) == null ? void 0 : _triggersArray$i10.saveResponse) === true) _this2.memoryCard.save(response);

                        if (!(((_triggersArray$i11 = triggersArray[i]) == null ? void 0 : _triggersArray$i11.shouldReturnPayload) === true)) {
                          _context11.next = 32;
                          break;
                        }

                        return _context11.abrupt("return", {
                          v: {
                            payload: payload,
                            response: response
                          }
                        });

                      case 32:
                        if (!(((_triggersArray$i12 = triggersArray[i]) == null ? void 0 : _triggersArray$i12.shouldReturnResponse) === true)) {
                          _context11.next = 34;
                          break;
                        }

                        return _context11.abrupt("return", {
                          v: response
                        });

                      case 34:
                        return _context11.abrupt("break", 73);

                      case 35:
                        if (triggersArray[i].pluginName) {
                          _context11.next = 37;
                          break;
                        }

                        throw new Error('callPluginEndpoint trigger does not have an plugin name to call');

                      case 37:
                        if (triggersArray[i].endpoint) {
                          _context11.next = 39;
                          break;
                        }

                        throw new Error('callPluginEndpoint trigger does not have an endpoint to call');

                      case 39:
                        if (!(((_triggersArray$i13 = triggersArray[i]) == null ? void 0 : _triggersArray$i13.shouldAwait) === true)) {
                          _context11.next = 44;
                          break;
                        }

                        if (triggersArray[i].useSavedResponse === true) {
                          triggersArray[i].params[triggersArray[i].savedResponseAs] = _this2.memoryCard.get(triggersArray[i].savedResponseAt);
                        }

                        _context11.next = 43;
                        return _regenerator.default.awrap((0, _Cloud.run)(triggersArray[i].pluginName + "_" + triggersArray[i].endpoint, triggersArray[i].params));

                      case 43:
                        response = _context11.sent;

                      case 44:
                        if (((_triggersArray$i14 = triggersArray[i]) == null ? void 0 : _triggersArray$i14.shouldAwait) === false) {
                          if (triggersArray[i].useSavedResponse === true) {
                            triggersArray[i].params[triggersArray[i].savedResponseAs] = _this2.memoryCard.get(triggersArray[i].savedResponseAt);
                          }

                          response = (0, _Cloud.run)(triggersArray[i].pluginName + "_" + triggersArray[i].endpoint, triggersArray[i].params);
                        }

                        if (!(triggersArray[i].runResponseTrigger === true)) {
                          _context11.next = 49;
                          break;
                        }

                        _context11.next = 48;
                        return _regenerator.default.awrap(_this2.handleTriggers(response.data.result.triggers, response.data.result.data));

                      case 48:
                        response = _context11.sent;

                      case 49:
                        if (((_triggersArray$i15 = triggersArray[i]) == null ? void 0 : _triggersArray$i15.saveResponse) === true) _this2.memoryCard.save(response);

                        if (!(((_triggersArray$i16 = triggersArray[i]) == null ? void 0 : _triggersArray$i16.runResponseTrigger) === false)) {
                          _context11.next = 52;
                          break;
                        }

                        return _context11.abrupt("return", "continue");

                      case 52:
                        if (!(((_triggersArray$i17 = triggersArray[i]) == null ? void 0 : _triggersArray$i17.shouldReturnPayload) === true)) {
                          _context11.next = 54;
                          break;
                        }

                        return _context11.abrupt("return", {
                          v: {
                            payload: 'payload',
                            response: response
                          }
                        });

                      case 54:
                        if (!(((_triggersArray$i18 = triggersArray[i]) == null ? void 0 : _triggersArray$i18.shouldReturnResponse) === true)) {
                          _context11.next = 56;
                          break;
                        }

                        return _context11.abrupt("return", {
                          v: response
                        });

                      case 56:
                        return _context11.abrupt("break", 73);

                      case 57:
                        if (_this2.ensureWeb3IsInstalled()) {
                          _context11.next = 59;
                          break;
                        }

                        throw new Error(ERROR_WEB3_MISSING);

                      case 59:
                        if (triggersArray[i].parameters) {
                          _context11.next = 61;
                          break;
                        }

                        throw new Error('web3SignV4 trigger does not have `parameters` to sign');

                      case 61:
                        if (triggersArray[i].from) {
                          _context11.next = 63;
                          break;
                        }

                        throw new Error('web3SignV4 trigger does not have a `from` address');

                      case 63:
                        if (!(((_triggersArray$i19 = triggersArray[i]) == null ? void 0 : _triggersArray$i19.shouldAwait) === true)) {
                          _context11.next = 66;
                          break;
                        }

                        _context11.next = 66;
                        return _regenerator.default.awrap(_this2.web3.currentProvider.send({
                          method: 'eth_signTypedData_v4',
                          params: triggersArray[i].parameters,
                          from: triggersArray[i].from
                        }, function (error, result) {
                          var _triggersArray$i20;

                          return _regenerator.default.async(function (_context9) {
                            while (1) {
                              switch (_context9.prev = _context9.next) {
                                case 0:
                                  if (!error) {
                                    _context9.next = 2;
                                    break;
                                  }

                                  throw new Error(error.message || error);

                                case 2:
                                  if (((_triggersArray$i20 = triggersArray[i]) == null ? void 0 : _triggersArray$i20.saveResponse) === true) _this2.memoryCard.save(result);
                                  response = result;

                                case 4:
                                case "end":
                                  return _context9.stop();
                              }
                            }
                          }, null, null, null, Promise);
                        }));

                      case 66:
                        if (((_triggersArray$i21 = triggersArray[i]) == null ? void 0 : _triggersArray$i21.shouldAwait) === false) {
                          _this2.web3.currentProvider.send({
                            method: 'eth_signTypedData_v4',
                            params: triggersArray[i].parameters,
                            from: triggersArray[i].from
                          }, function (error, result) {
                            var _triggersArray$i22;

                            return _regenerator.default.async(function (_context10) {
                              while (1) {
                                switch (_context10.prev = _context10.next) {
                                  case 0:
                                    if (!error) {
                                      _context10.next = 2;
                                      break;
                                    }

                                    throw new Error(error.message || error);

                                  case 2:
                                    if (((_triggersArray$i22 = triggersArray[i]) == null ? void 0 : _triggersArray$i22.saveResponse) === true) _this2.memoryCard.save(result);
                                    response = result;

                                  case 4:
                                  case "end":
                                    return _context10.stop();
                                }
                              }
                            }, null, null, null, Promise);
                          });
                        }

                        if (!(((_triggersArray$i23 = triggersArray[i]) == null ? void 0 : _triggersArray$i23.shouldReturnPayload) === true)) {
                          _context11.next = 69;
                          break;
                        }

                        return _context11.abrupt("return", {
                          v: {
                            payload: payload,
                            response: response
                          }
                        });

                      case 69:
                        if (!(((_triggersArray$i24 = triggersArray[i]) == null ? void 0 : _triggersArray$i24.shouldReturnResponse) === true)) {
                          _context11.next = 71;
                          break;
                        }

                        return _context11.abrupt("return", {
                          v: response
                        });

                      case 71:
                        return _context11.abrupt("break", 73);

                      case 72:
                        throw new Error("Unknown trigger: \"" + ((_triggersArray$i25 = triggersArray[i]) == null ? void 0 : _triggersArray$i25.name) + "\"");

                      case 73:
                      case "end":
                        return _context11.stop();
                    }
                  }
                }, null, null, null, Promise);
              };

              i = 0;

            case 4:
              if (!(i < triggersArray.length)) {
                _context12.next = 15;
                break;
              }

              _context12.next = 7;
              return _regenerator.default.awrap(_loop(i));

            case 7:
              _ret = _context12.sent;

              if (!(_ret === "continue")) {
                _context12.next = 10;
                break;
              }

              return _context12.abrupt("continue", 12);

            case 10:
              if (!(typeof _ret === "object")) {
                _context12.next = 12;
                break;
              }

              return _context12.abrupt("return", _ret.v);

            case 12:
              i++;
              _context12.next = 4;
              break;

            case 15:
              this.memoryCard.deleteSaved();

            case 16:
            case "end":
              return _context12.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "getAllERC20",
    value: function () {
      var _ref,
          chain,
          address,
          result,
          _args13 = arguments;

      return _regenerator.default.async(function (_context13) {
        while (1) {
          switch (_context13.prev = _context13.next) {
            case 0:
              _ref = _args13.length > 0 && _args13[0] !== undefined ? _args13[0] : {}, chain = _ref.chain, address = _ref.address;
              _context13.next = 3;
              return _regenerator.default.awrap((0, _Cloud.run)('getAllERC20', {
                chain: chain,
                address: address
              }));

            case 3:
              result = _context13.sent;
              return _context13.abrupt("return", result);

            case 5:
            case "end":
              return _context13.stop();
          }
        }
      }, null, null, null, Promise);
    }
  }, {
    key: "getERC20",
    value: function () {
      var _ref2,
          chain,
          address,
          symbol,
          tokenAddress,
          result,
          _args14 = arguments;

      return _regenerator.default.async(function (_context14) {
        while (1) {
          switch (_context14.prev = _context14.next) {
            case 0:
              _ref2 = _args14.length > 0 && _args14[0] !== undefined ? _args14[0] : {}, chain = _ref2.chain, address = _ref2.address, symbol = _ref2.symbol, tokenAddress = _ref2.tokenAddress;
              result = (0, _Cloud.run)('getERC20', {
                chain: chain,
                address: address,
                symbol: symbol,
                tokenAddress: tokenAddress
              });
              return _context14.abrupt("return", result);

            case 3:
            case "end":
              return _context14.stop();
          }
        }
      }, null, null, null, Promise);
    }
  }, {
    key: "getNFTs",
    value: function () {
      var _ref3 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref3$chain = _ref3.chain,
          chain = _ref3$chain === void 0 ? 'Eth' : _ref3$chain,
          _ref3$address = _ref3.address,
          address = _ref3$address === void 0 ? '' : _ref3$address;

      return (0, _Cloud.run)('getNFTs_old', {
        chain: chain,
        address: address
      });
    }
  }, {
    key: "getNFTsCount",
    value: function () {
      var _ref4 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref4$chain = _ref4.chain,
          chain = _ref4$chain === void 0 ? 'Eth' : _ref4$chain,
          _ref4$address = _ref4.address,
          address = _ref4$address === void 0 ? '' : _ref4$address;

      return (0, _Cloud.run)('getNFTsCount_old', {
        chain: chain,
        address: address
      });
    }
  }, {
    key: "getTransactions",
    value: function () {
      var _ref5 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref5$chain = _ref5.chain,
          chain = _ref5$chain === void 0 ? 'Eth' : _ref5$chain,
          _ref5$address = _ref5.address,
          address = _ref5$address === void 0 ? '' : _ref5$address,
          _ref5$order = _ref5.order,
          order = _ref5$order === void 0 ? 'desc' : _ref5$order;

      return (0, _Cloud.run)('getTransactions', {
        chain: chain,
        address: address,
        order: order
      });
    }
  }, {
    key: "getTransactionsCount",
    value: function () {
      var _ref6 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref6$chain = _ref6.chain,
          chain = _ref6$chain === void 0 ? 'Eth' : _ref6$chain,
          _ref6$address = _ref6.address,
          address = _ref6$address === void 0 ? '' : _ref6$address;

      return (0, _Cloud.run)('getTransactionsCount', {
        chain: chain,
        address: address
      });
    }
  }, {
    key: "transfer",
    value: function () {
      var _ref7,
          _ref7$type,
          type,
          _ref7$receiver,
          receiver,
          _ref7$contractAddress,
          contractAddress,
          contract_address,
          _ref7$amount,
          amount,
          _ref7$tokenId,
          tokenId,
          token_id,
          _ref7$system,
          system,
          _ref7$awaitReceipt,
          awaitReceipt,
          options,
          web3,
          sender,
          transferOperation,
          customToken,
          transferEvents,
          _args15 = arguments;

      return _regenerator.default.async(function (_context15) {
        while (1) {
          switch (_context15.prev = _context15.next) {
            case 0:
              _ref7 = _args15.length > 0 && _args15[0] !== undefined ? _args15[0] : {}, _ref7$type = _ref7.type, type = _ref7$type === void 0 ? 'native' : _ref7$type, _ref7$receiver = _ref7.receiver, receiver = _ref7$receiver === void 0 ? '' : _ref7$receiver, _ref7$contractAddress = _ref7.contractAddress, contractAddress = _ref7$contractAddress === void 0 ? '' : _ref7$contractAddress, contract_address = _ref7.contract_address, _ref7$amount = _ref7.amount, amount = _ref7$amount === void 0 ? '' : _ref7$amount, _ref7$tokenId = _ref7.tokenId, tokenId = _ref7$tokenId === void 0 ? '' : _ref7$tokenId, token_id = _ref7.token_id, _ref7$system = _ref7.system, system = _ref7$system === void 0 ? 'evm' : _ref7$system, _ref7$awaitReceipt = _ref7.awaitReceipt, awaitReceipt = _ref7$awaitReceipt === void 0 ? true : _ref7$awaitReceipt;
              contractAddress = contractAddress || contract_address;
              tokenId = tokenId || token_id;
              options = {
                receiver: receiver,
                contractAddress: contractAddress,
                amount: amount,
                tokenId: tokenId,
                system: system,
                awaitReceipt: awaitReceipt
              };

              _TransferUtils.default.isSupportedType(type);

              _TransferUtils.default.validateInput(type, options);

              if (this.ensureWeb3IsInstalled()) {
                _context15.next = 8;
                break;
              }

              throw new Error(ERROR_WEB3_MISSING);

            case 8:
              web3 = this.web3;
              _context15.t0 = _regenerator.default;
              _context15.next = 12;
              return _regenerator.default.awrap(web3.eth.getAccounts());

            case 12:
              _context15.t1 = _context15.sent[0];
              _context15.next = 15;
              return _context15.t0.awrap.call(_context15.t0, _context15.t1);

            case 15:
              sender = _context15.sent;

              if (sender) {
                _context15.next = 18;
                break;
              }

              throw new Error('Sender address not found');

            case 18:
              if (tokenId) _TransferUtils.default.isUint256(tokenId);
              if (type !== 'native') customToken = new web3.eth.Contract(_TransferUtils.default.abi[type], contractAddress);
              _context15.t2 = type;
              _context15.next = _context15.t2 === 'native' ? 23 : _context15.t2 === 'erc20' ? 25 : _context15.t2 === 'erc721' ? 27 : _context15.t2 === 'erc1155' ? 29 : 31;
              break;

            case 23:
              transferOperation = web3.eth.sendTransaction({
                from: sender,
                to: receiver,
                value: amount
              });
              return _context15.abrupt("break", 32);

            case 25:
              transferOperation = customToken.methods.transfer(receiver, amount).send({
                from: sender
              });
              return _context15.abrupt("break", 32);

            case 27:
              transferOperation = customToken.methods.safeTransferFrom(sender, receiver, "" + tokenId).send({
                from: sender
              });
              return _context15.abrupt("break", 32);

            case 29:
              transferOperation = customToken.methods.safeTransferFrom(sender, receiver, "" + tokenId, amount, '0x').send({
                from: sender
              });
              return _context15.abrupt("break", 32);

            case 31:
              throw new Error("Unknown transfer type: \"" + type + "\"");

            case 32:
              if (!awaitReceipt) {
                _context15.next = 34;
                break;
              }

              return _context15.abrupt("return", transferOperation);

            case 34:
              transferEvents = new EventEmitter();
              transferOperation.on('transactionHash', function (hash) {
                transferEvents.emit('transactionHash', hash);
              }).on('receipt', function (receipt) {
                transferEvents.emit('receipt', receipt);
              }).on('confirmation', function (confirmationNumber, receipt) {
                transferEvents.emit('confirmation', (confirmationNumber, receipt));
              }).on('error', function (error) {
                transferEvents.emit('error', error);
                throw error;
              });
              return _context15.abrupt("return", transferEvents);

            case 37:
            case "end":
              return _context15.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "executeFunction",
    value: function () {
      var _ref8,
          contractAddress,
          abi,
          functionName,
          msgValue,
          _ref8$awaitReceipt,
          awaitReceipt,
          _ref8$params,
          params,
          web3,
          contractOptions,
          functionData,
          stateMutability,
          isReadFunction,
          currentAddress,
          errors,
          _iterator,
          _step,
          input,
          value,
          parsedInputs,
          contract,
          customFunction,
          response,
          contractExecuteEvents,
          _args16 = arguments;

      return _regenerator.default.async(function (_context16) {
        while (1) {
          switch (_context16.prev = _context16.next) {
            case 0:
              _ref8 = _args16.length > 0 && _args16[0] !== undefined ? _args16[0] : {}, contractAddress = _ref8.contractAddress, abi = _ref8.abi, functionName = _ref8.functionName, msgValue = _ref8.msgValue, _ref8$awaitReceipt = _ref8.awaitReceipt, awaitReceipt = _ref8$awaitReceipt === void 0 ? true : _ref8$awaitReceipt, _ref8$params = _ref8.params, params = _ref8$params === void 0 ? {} : _ref8$params;

              if (this.ensureWeb3IsInstalled()) {
                _context16.next = 3;
                break;
              }

              throw new Error(ERROR_WEB3_MISSING);

            case 3:
              web3 = this.web3;
              contractOptions = {};
              functionData = abi.find(function (x) {
                return x.name === functionName;
              });

              if (functionData) {
                _context16.next = 8;
                break;
              }

              throw new Error('Function does not exist in abi');

            case 8:
              stateMutability = functionData == null ? void 0 : functionData.stateMutability;
              isReadFunction = stateMutability === 'view' || stateMutability === 'pure';

              if (isReadFunction) {
                _context16.next = 22;
                break;
              }

              if (params.from) {
                _context16.next = 22;
                break;
              }

              _context16.t0 = _regenerator.default;
              _context16.next = 15;
              return _regenerator.default.awrap(web3.eth.getAccounts());

            case 15:
              _context16.t1 = _context16.sent[0];
              _context16.next = 18;
              return _context16.t0.awrap.call(_context16.t0, _context16.t1);

            case 18:
              currentAddress = _context16.sent;

              if (currentAddress) {
                _context16.next = 21;
                break;
              }

              throw new Error('From address is required');

            case 21:
              contractOptions.from = currentAddress;

            case 22:
              errors = [];

              for (_iterator = _createForOfIteratorHelperLoose(functionData.inputs); !(_step = _iterator()).done;) {
                input = _step.value;
                value = params[input.name];

                if (!value && typeof value !== 'number' && typeof value !== 'boolean') {
                  errors.push(input.name + " is required");
                }
              }

              if (!(errors.length > 0)) {
                _context16.next = 26;
                break;
              }

              throw errors;

            case 26:
              parsedInputs = functionData.inputs.map(function (x) {
                return params[x.name];
              });
              contract = new web3.eth.Contract(abi, contractAddress, contractOptions);
              customFunction = contract.methods[functionName];
              response = isReadFunction ? customFunction.apply(void 0, (0, _toConsumableArray2.default)(Object.values(parsedInputs))).call() : customFunction.apply(void 0, (0, _toConsumableArray2.default)(Object.values(parsedInputs))).send(msgValue ? {
                value: msgValue
              } : null);

              if (!awaitReceipt) {
                _context16.next = 32;
                break;
              }

              return _context16.abrupt("return", response);

            case 32:
              contractExecuteEvents = new EventEmitter();
              response.on('transactionHash', function (hash) {
                contractExecuteEvents.emit('transactionHash', hash);
              }).on('receipt', function (receipt) {
                contractExecuteEvents.emit('receipt', receipt);
              }).on('confirmation', function (confirmationNumber, receipt) {
                contractExecuteEvents.emit('confirmation', (confirmationNumber, receipt));
              }).on('error', function (error) {
                contractExecuteEvents.emit('error', error);
                throw error;
              });
              return _context16.abrupt("return", contractExecuteEvents);

            case 35:
            case "end":
              return _context16.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "getSigningData",
    value: function () {
      return "Moralis Authentication";
    }
  }, {
    key: "on",
    value: function (eventName, cb) {
      var _window = window,
          ethereum = _window.ethereum;

      if (!ethereum || !ethereum.on) {
        console.warn(WARNING);
        return function () {
          console.warn(WARNING);
        };
      }

      ethereum.on(eventName, cb);
      return function () {
        console.warn('UNSUB NOT SUPPORTED');
      };
    }
  }, {
    key: "getChainId",
    value: function () {
      return _regenerator.default.async(function (_context17) {
        while (1) {
          switch (_context17.prev = _context17.next) {
            case 0:
              if (!this.ensureWeb3IsInstalled()) {
                _context17.next = 4;
                break;
              }

              _context17.next = 3;
              return _regenerator.default.awrap(this.web3.eth.net.getId());

            case 3:
              return _context17.abrupt("return", _context17.sent);

            case 4:
              throw new Error(ERROR_WEB3_MISSING);

            case 5:
            case "end":
              return _context17.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "ensureWeb3IsInstalled",
    value: function () {
      return this.web3 ? true : false;
    }
  }, {
    key: "isMetaMaskInstalled",
    value: function () {
      return _regenerator.default.async(function (_context18) {
        while (1) {
          switch (_context18.prev = _context18.next) {
            case 0:
              _context18.next = 2;
              return _regenerator.default.awrap((0, _detectProvider.default)());

            case 2:
              if (!_context18.sent) {
                _context18.next = 6;
                break;
              }

              _context18.t0 = true;
              _context18.next = 7;
              break;

            case 6:
              _context18.t0 = false;

            case 7:
              return _context18.abrupt("return", _context18.t0);

            case 8:
            case "end":
              return _context18.stop();
          }
        }
      }, null, null, null, Promise);
    }
  }, {
    key: "switchNetwork",
    value: function (chainId) {
      var currentNetwork;
      return _regenerator.default.async(function (_context19) {
        while (1) {
          switch (_context19.prev = _context19.next) {
            case 0:
              chainId = verifyChainId(chainId);
              _context19.t0 = fromDecimalToHex;
              _context19.next = 4;
              return _regenerator.default.awrap(this.getChainId());

            case 4:
              _context19.t1 = _context19.sent;
              currentNetwork = (0, _context19.t0)(_context19.t1);

              if (!(currentNetwork === chainId)) {
                _context19.next = 8;
                break;
              }

              return _context19.abrupt("return");

            case 8:
              _context19.next = 10;
              return _regenerator.default.awrap(window.ethereum.request({
                method: 'wallet_switchEthereumChain',
                params: [{
                  chainId: chainId
                }]
              }));

            case 10:
            case "end":
              return _context19.stop();
          }
        }
      }, null, this, null, Promise);
    }
  }, {
    key: "addNetwork",
    value: function (chainId, chainName, currencyName, currencySymbol, rpcUrl, blockExplorerUrl) {
      return _regenerator.default.async(function (_context20) {
        while (1) {
          switch (_context20.prev = _context20.next) {
            case 0:
              chainId = verifyChainId(chainId);
              _context20.next = 3;
              return _regenerator.default.awrap(window.ethereum.request({
                method: 'wallet_addEthereumChain',
                params: [{
                  chainId: chainId,
                  chainName: chainName,
                  nativeCurrency: {
                    name: currencyName,
                    symbol: currencySymbol,
                    decimals: 18
                  },
                  rpcUrls: [rpcUrl],
                  blockExplorerUrls: [blockExplorerUrl]
                }]
              }));

            case 3:
            case "end":
              return _context20.stop();
          }
        }
      }, null, null, null, Promise);
    }
  }]);
  return MoralisWeb3;
}();

MoralisWeb3.memoryCard = {
  save: function (what) {
    this.saved = what;
  },
  get: function (where) {
    if (!this.saved) throw new Error('Nothing saved to memory card');
    if (where.length === 0) return this.getSaved();
    var tmp;
    var savedTmp = this.saved;

    for (var i = 0; i < where.length; i++) {
      tmp = savedTmp[where[i]];
      savedTmp = tmp;
    }

    return savedTmp;
  },
  getSaved: function () {
    return this.saved;
  },
  deleteSaved: function () {
    this.saved = undefined;
  }
};

function fromDecimalToHex(number) {
  if (typeof number !== 'number') throw 'The input provided should be a number';
  return "0x" + number.toString(16);
}

function verifyChainId(chainId) {
  if (typeof chainId === 'number') chainId = fromDecimalToHex(chainId);
  return chainId;
}

MoralisWeb3.onConnect = MoralisWeb3.on.bind(MoralisWeb3, EthereumEvents.CONNECT);
MoralisWeb3.onDisconnect = MoralisWeb3.on.bind(MoralisWeb3, EthereumEvents.DISCONNECT);
MoralisWeb3.onChainChanged = MoralisWeb3.on.bind(MoralisWeb3, EthereumEvents.CHAIN_CHANGED);
MoralisWeb3.onAccountsChanged = MoralisWeb3.on.bind(MoralisWeb3, EthereumEvents.ACCOUNTS_CHANGED);
var _default = MoralisWeb3;
exports.default = _default;
