// moralis/lib/weapp/MoralisWeb3.js
"use strict";

var _sliceInstanceProperty = require("@babel/runtime-corejs3/core-js-stable/instance/slice");

var _Array$from = require("@babel/runtime-corejs3/core-js-stable/array/from");

var _Symbol = require("@babel/runtime-corejs3/core-js-stable/symbol");

var _getIteratorMethod = require("@babel/runtime-corejs3/core-js/get-iterator-method");

var _Array$isArray = require("@babel/runtime-corejs3/core-js-stable/array/is-array");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = exports.EthereumEvents = void 0;

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _stringify = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/json/stringify"));

var _find = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/find"));

var _values = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/values"));

var _bind = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/bind"));

var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs3/regenerator"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/toConsumableArray"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/typeof"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncToGenerator"));

var _construct2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/construct"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/defineProperty"));

var _web = _interopRequireDefault(require("web3"));

var _ParseObject = _interopRequireDefault(require("./ParseObject"));

var _ParseQuery = _interopRequireDefault(require("./ParseQuery"));

var _ParseUser = _interopRequireDefault(require("./ParseUser"));

var _ParseACL = _interopRequireDefault(require("./ParseACL"));

var _MoralisErd = _interopRequireDefault(require("./MoralisErd"));

var _MoralisDot = _interopRequireDefault(require("./MoralisDot"));

var _MoralisSol = _interopRequireDefault(require("./MoralisSol"));

var _MoralisWalletConnectProvider = _interopRequireDefault(require("./MoralisWalletConnectProvider"));

var _MoralisCustomProvider = _interopRequireDefault(require("./MoralisCustomProvider"));

var _MoralisInjectedProvider = _interopRequireDefault(require("./MoralisInjectedProvider"));

var _TransferUtils = _interopRequireDefault(require("./TransferUtils"));

var _Cloud = require("./Cloud");

var _detectProvider = _interopRequireDefault(require("@metamask/detect-provider"));

var _createSigningData = _interopRequireDefault(require("./createSigningData"));

var _context26, _context27, _context28, _context29;

function _createForOfIteratorHelper(o, allowArrayLike) {
  var it = typeof _Symbol !== "undefined" && _getIteratorMethod(o) || o["@@iterator"];

  if (!it) {
    if (_Array$isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
      if (it) o = it;
      var i = 0;

      var F = function () {};

      return {
        s: F,
        n: function () {
          if (i >= o.length) return {
            done: true
          };
          return {
            done: false,
            value: o[i++]
          };
        },
        e: function (_e) {
          throw _e;
        },
        f: F
      };
    }

    throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
  }

  var normalCompletion = true,
      didErr = false,
      err;
  return {
    s: function () {
      it = it.call(o);
    },
    n: function () {
      var step = it.next();
      normalCompletion = step.done;
      return step;
    },
    e: function (_e2) {
      didErr = true;
      err = _e2;
    },
    f: function () {
      try {
        if (!normalCompletion && it.return != null) it.return();
      } finally {
        if (didErr) throw err;
      }
    }
  };
}

function _unsupportedIterableToArray(o, minLen) {
  var _context30;

  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);

  var n = _sliceInstanceProperty(_context30 = Object.prototype.toString.call(o)).call(_context30, 8, -1);

  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return _Array$from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) {
    arr2[i] = arr[i];
  }

  return arr2;
}

var EventEmitter = require('events');

var EthereumEvents = {
  CONNECT: 'connect',
  DISCONNECT: 'disconnect',
  ACCOUNTS_CHANGED: 'accountsChanged',
  CHAIN_CHANGED: 'chainChanged'
};
exports.EthereumEvents = EthereumEvents;
var WARNING = 'Non ethereum enabled browser';
var ERROR_WEB3_MISSING = 'Missing web3 instance, make sure to call Moralis.enableWeb3() or Moralis.authenticate()';

var MoralisWeb3 = /*#__PURE__*/function () {
  function MoralisWeb3() {
    (0, _classCallCheck2.default)(this, MoralisWeb3);
    var MWeb3 = typeof _web.default === 'function' ? _web.default : window.Web3;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return (0, _construct2.default)(MWeb3, args);
  }

  (0, _createClass2.default)(MoralisWeb3, null, [{
    key: "isWeb3Enabled",
    value: function () {
      return this.ensureWeb3IsInstalled();
    }
  }, {
    key: "setEnableWeb3",
    value: function (fn) {
      this.customEnableWeb3 = fn;
    }
  }, {
    key: "enableWeb3",
    value: function () {
      var _enableWeb = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(options) {
        var web3, Web3Provider, web3Provider;
        return _regenerator.default.wrap(function (_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this.customEnableWeb3) {
                  _context.next = 6;
                  break;
                }

                _context.next = 3;
                return this.customEnableWeb3(options);

              case 3:
                web3 = _context.sent;
                _context.next = 13;
                break;

              case 6:
                if (this.speedyNodeApiKey) {
                  options.speedyNodeApiKey = this.speedyNodeApiKey;
                  options.provider = 'custom';
                }

                Web3Provider = MoralisWeb3.getWeb3Provider(options);
                web3Provider = new Web3Provider();
                this.activeWeb3Provider = web3Provider;
                _context.next = 12;
                return web3Provider.activate(options);

              case 12:
                web3 = _context.sent;

              case 13:
                this.web3 = web3;
                return _context.abrupt("return", web3);

              case 15:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      return function () {
        return _enableWeb.apply(this, arguments);
      };
    }()
  }, {
    key: "enable",
    value: function () {
      var _enable = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(options) {
        return _regenerator.default.wrap(function (_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                // eslint-disable-next-line no-console
                console.warn('Moralis.enable() is deprecated and will be removed, use Moralis.enableWeb3() instead.');
                return _context2.abrupt("return", this.enableWeb3(options));

              case 2:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      return function () {
        return _enable.apply(this, arguments);
      };
    }()
  }, {
    key: "isDotAuth",
    value: function (options) {
      switch (options === null || options === void 0 ? void 0 : options.type) {
        case 'dot':
        case 'polkadot':
        case 'kusama':
          return true;

        default:
          return false;
      }
    }
  }, {
    key: "isElrondAuth",
    value: function (options) {
      switch (options === null || options === void 0 ? void 0 : options.type) {
        case 'erd':
        case 'elrond':
          return true;

        default:
          return false;
      }
    }
  }, {
    key: "isSolAuth",
    value: function (options) {
      switch (options === null || options === void 0 ? void 0 : options.type) {
        case 'sol':
          return true;

        default:
          return false;
      }
    }
  }, {
    key: "getWeb3Provider",
    value: function (options) {
      switch (options === null || options === void 0 ? void 0 : options.provider) {
        case 'walletconnect':
        case 'walletConnect':
        case 'wc':
          return _MoralisWalletConnectProvider.default;

        case 'custom':
          return _MoralisCustomProvider.default;

        default:
          return _MoralisInjectedProvider.default;
      }
    }
  }, {
    key: "cleanup",
    value: function () {
      var _cleanup = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee3() {
        return _regenerator.default.wrap(function (_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                if (!this.activeWeb3Provider) {
                  _context3.next = 3;
                  break;
                }

                _context3.next = 3;
                return this.activeWeb3Provider.deactivate();

              case 3:
                // Prevent a bug when there is stale data active
                _MoralisWalletConnectProvider.default.cleanupStaleData();

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      return function () {
        return _cleanup.apply(this, arguments);
      };
    }()
  }, {
    key: "authenticate",
    value: function () {
      var _authenticate = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee4(options) {
        var isLoggedIn, web3, message, data, accounts, accountsLower, _accountsLower, ethAddress, signature, authData, user;

        return _regenerator.default.wrap(function (_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return _ParseUser.default.currentAsync();

              case 2:
                isLoggedIn = _context4.sent;

                if (!isLoggedIn) {
                  _context4.next = 6;
                  break;
                }

                _context4.next = 6;
                return _ParseUser.default.logOut();

              case 6:
                _context4.next = 8;
                return MoralisWeb3.cleanup();

              case 8:
                if (!MoralisWeb3.isDotAuth(options)) {
                  _context4.next = 10;
                  break;
                }

                return _context4.abrupt("return", _MoralisDot.default.authenticate(options));

              case 10:
                if (!MoralisWeb3.isElrondAuth(options)) {
                  _context4.next = 12;
                  break;
                }

                return _context4.abrupt("return", _MoralisErd.default.authenticate(options));

              case 12:
                if (!MoralisWeb3.isSolAuth(options)) {
                  _context4.next = 14;
                  break;
                }

                return _context4.abrupt("return", _MoralisSol.default.authenticate(options));

              case 14:
                _context4.next = 16;
                return this.enableWeb3(options);

              case 16:
                web3 = _context4.sent;
                message = (options === null || options === void 0 ? void 0 : options.signingMessage) || MoralisWeb3.getSigningData();
                _context4.next = 20;
                return (0, _createSigningData.default)(message);

              case 20:
                data = _context4.sent;
                _context4.next = 23;
                return web3.eth.getAccounts();

              case 23:
                accounts = _context4.sent;
                accountsLower = (0, _map.default)(accounts).call(accounts, function (v) {
                  return v.toLowerCase();
                });
                _accountsLower = (0, _slicedToArray2.default)(accountsLower, 1), ethAddress = _accountsLower[0];

                if (ethAddress) {
                  _context4.next = 28;
                  break;
                }

                throw new Error('Address not found');

              case 28:
                _context4.next = 30;
                return web3.eth.personal.sign(data, ethAddress, '');

              case 30:
                signature = _context4.sent;

                if (signature) {
                  _context4.next = 33;
                  break;
                }

                throw new Error('Data not signed');

              case 33:
                authData = {
                  id: ethAddress,
                  signature: signature,
                  data: data
                };
                _context4.next = 36;
                return _ParseUser.default.logInWith('moralisEth', {
                  authData: authData
                });

              case 36:
                user = _context4.sent;

                if (user) {
                  _context4.next = 39;
                  break;
                }

                throw new Error('Could not get user');

              case 39:
                _context4.next = 41;
                return user.setACL(new _ParseACL.default(user));

              case 41:
                user.addAllUnique('accounts', accountsLower);
                user.set('ethAddress', ethAddress);
                _context4.next = 45;
                return user.save(null, options);

              case 45:
                return _context4.abrupt("return", user);

              case 46:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      return function () {
        return _authenticate.apply(this, arguments);
      };
    }()
  }, {
    key: "link",
    value: function () {
      var _link = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee5(account, options) {
        var web3, data, user, ethAddress, EthAddress, query, ethAddressRecord, signature, authData;
        return _regenerator.default.wrap(function (_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return MoralisWeb3.enableWeb3(options);

              case 2:
                web3 = _context5.sent;
                data = (options === null || options === void 0 ? void 0 : options.signingMessage) || MoralisWeb3.getSigningData();
                _context5.next = 6;
                return _ParseUser.default.currentAsync();

              case 6:
                user = _context5.sent;
                ethAddress = account.toLowerCase();
                EthAddress = _ParseObject.default.extend('_EthAddress');
                query = new _ParseQuery.default(EthAddress);
                _context5.next = 12;
                return query.get(ethAddress).catch(function () {
                  return null;
                });

              case 12:
                ethAddressRecord = _context5.sent;

                if (ethAddressRecord) {
                  _context5.next = 20;
                  break;
                }

                _context5.next = 16;
                return web3.eth.personal.sign(data, account, '');

              case 16:
                signature = _context5.sent;
                authData = {
                  id: ethAddress,
                  signature: signature,
                  data: data
                };
                _context5.next = 20;
                return user.linkWith('moralisEth', {
                  authData: authData
                });

              case 20:
                user.addAllUnique('accounts', [ethAddress]);
                user.set('ethAddress', ethAddress);
                _context5.next = 24;
                return user.save(null, options);

              case 24:
                return _context5.abrupt("return", user);

              case 25:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5);
      }));

      return function () {
        return _link.apply(this, arguments);
      };
    }()
  }, {
    key: "unlink",
    value: function () {
      var _unlink = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee6(account) {
        var _user$get;

        var accountsLower, EthAddress, query, ethAddressRecord, user, accounts, nextAccounts;
        return _regenerator.default.wrap(function (_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                accountsLower = account.toLowerCase();
                EthAddress = _ParseObject.default.extend('_EthAddress');
                query = new _ParseQuery.default(EthAddress);
                _context6.next = 5;
                return query.get(accountsLower);

              case 5:
                ethAddressRecord = _context6.sent;
                _context6.next = 8;
                return ethAddressRecord.destroy();

              case 8:
                _context6.next = 10;
                return _ParseUser.default.currentAsync();

              case 10:
                user = _context6.sent;
                accounts = (_user$get = user.get('accounts')) !== null && _user$get !== void 0 ? _user$get : [];
                nextAccounts = (0, _filter.default)(accounts).call(accounts, function (v) {
                  return v !== accountsLower;
                });
                user.set('accounts', nextAccounts);
                user.set('ethAddress', nextAccounts[0]);
                _context6.next = 17;
                return user._unlinkFrom('moralisEth');

              case 17:
                _context6.next = 19;
                return user.save();

              case 19:
                return _context6.abrupt("return", user);

              case 20:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6);
      }));

      return function () {
        return _unlink.apply(this, arguments);
      };
    }()
  }, {
    key: "initPlugins",
    value: function () {
      var _initPlugins = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee8(installedPlugins) {
        var _this = this;

        var specs, allPlugins;
        return _regenerator.default.wrap(function (_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.t0 = installedPlugins;

                if (_context10.t0) {
                  _context10.next = 5;
                  break;
                }

                _context10.next = 4;
                return (0, _Cloud.run)('getPluginSpecs');

              case 4:
                _context10.t0 = _context10.sent;

              case 5:
                specs = _context10.t0;
                if (!this.Plugins) this.Plugins = {};

                if (specs) {
                  _context10.next = 9;
                  break;
                }

                return _context10.abrupt("return");

              case 9:
                allPlugins = this.Plugins;
                (0, _forEach.default)(specs).call(specs, function (plugin) {
                  var _context7;

                  allPlugins[plugin.name] = {};
                  (0, _forEach.default)(_context7 = plugin.functions).call(_context7, function (f) {
                    allPlugins[plugin.name][f] = /*#__PURE__*/function () {
                      var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee7(params, options) {
                        var _context8;

                        var response, error, triggerReturn;
                        return _regenerator.default.wrap(function (_context9) {
                          while (1) {
                            switch (_context9.prev = _context9.next) {
                              case 0:
                                if (!options) options = {};
                                _context9.next = 3;
                                return (0, _Cloud.run)((0, _concat.default)(_context8 = "".concat(plugin.name, "_")).call(_context8, f), params);

                              case 3:
                                response = _context9.sent;

                                if (response.data.success) {
                                  _context9.next = 7;
                                  break;
                                }

                                error = (0, _stringify.default)(response.data.data, null, 2);
                                throw new Error("Something went wrong\n".concat(error));

                              case 7:
                                if (!(options.disableTriggers !== true)) {
                                  _context9.next = 13;
                                  break;
                                }

                                _context9.next = 10;
                                return _this.handleTriggers(response.data.result.triggers, response.data.result.data);

                              case 10:
                                triggerReturn = _context9.sent;

                                if (!triggerReturn) {
                                  _context9.next = 13;
                                  break;
                                }

                                return _context9.abrupt("return", triggerReturn);

                              case 13:
                                return _context9.abrupt("return", response.data.result);

                              case 14:
                              case "end":
                                return _context9.stop();
                            }
                          }
                        }, _callee7);
                      }));

                      return function () {
                        return _ref.apply(this, arguments);
                      };
                    }();
                  });
                });
                this.Plugins = allPlugins;

              case 12:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee8, this);
      }));

      return function () {
        return _initPlugins.apply(this, arguments);
      };
    }()
  }, {
    key: "handleTriggers",
    value: function () {
      var _handleTriggers = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee11(triggersArray, payload) {
        var _this2 = this;

        var response, _loop, i, _ret;

        return _regenerator.default.wrap(function (_context16) {
          while (1) {
            switch (_context16.prev = _context16.next) {
              case 0:
                if (triggersArray) {
                  _context16.next = 2;
                  break;
                }

                return _context16.abrupt("return");

              case 2:
                _loop = /*#__PURE__*/_regenerator.default.mark(function _loop(i) {
                  var _triggersArray$i, _triggersArray$i2, _triggersArray$i3, _triggersArray$i4, _triggersArray$i5, _triggersArray$i6, _triggersArray$i7, _triggersArray$i8, _triggersArray$i9, _triggersArray$i10, _triggersArray$i11, _triggersArray$i12, _triggersArray$i13, _triggersArray$i14, _triggersArray$i15, _triggersArray$i16, _triggersArray$i17, _triggersArray$i18, _triggersArray$i19, _triggersArray$i21, _triggersArray$i23, _triggersArray$i24, _triggersArray$i25;

                  var _context11, _context12;

                  return _regenerator.default.wrap(function (_context15) {
                    while (1) {
                      switch (_context15.prev = _context15.next) {
                        case 0:
                          _context15.t0 = triggersArray[i].name;
                          _context15.next = _context15.t0 === 'openUrl' ? 3 : _context15.t0 === 'web3Transaction' ? 5 : _context15.t0 === 'web3Sign' ? 18 : _context15.t0 === 'callPluginEndpoint' ? 35 : _context15.t0 === 'web3SignV4' ? 57 : 72;
                          break;

                        case 3:
                          if (triggersArray[i].newTab) window.open(triggersArray[i].url);else window.open(triggersArray[i].url, '_self');
                          return _context15.abrupt("break", 73);

                        case 5:
                          if (_this2.ensureWeb3IsInstalled()) {
                            _context15.next = 7;
                            break;
                          }

                          throw new Error(ERROR_WEB3_MISSING);

                        case 7:
                          if (!(((_triggersArray$i = triggersArray[i]) === null || _triggersArray$i === void 0 ? void 0 : _triggersArray$i.shouldAwait) === true)) {
                            _context15.next = 11;
                            break;
                          }

                          _context15.next = 10;
                          return _this2.web3.eth.sendTransaction((_triggersArray$i2 = triggersArray[i]) === null || _triggersArray$i2 === void 0 ? void 0 : _triggersArray$i2.data);

                        case 10:
                          response = _context15.sent;

                        case 11:
                          // Trigger a web3 transaction (does NOT await)
                          if (((_triggersArray$i3 = triggersArray[i]) === null || _triggersArray$i3 === void 0 ? void 0 : _triggersArray$i3.shouldAwait) === false) response = _this2.web3.eth.sendTransaction((_triggersArray$i4 = triggersArray[i]) === null || _triggersArray$i4 === void 0 ? void 0 : _triggersArray$i4.data); // Save the response returned by the web3 trasanction

                          if (((_triggersArray$i5 = triggersArray[i]) === null || _triggersArray$i5 === void 0 ? void 0 : _triggersArray$i5.saveResponse) === true) _this2.memoryCard.save(response); // Return payload and response

                          if (!(((_triggersArray$i6 = triggersArray[i]) === null || _triggersArray$i6 === void 0 ? void 0 : _triggersArray$i6.shouldReturnPayload) === true)) {
                            _context15.next = 15;
                            break;
                          }

                          return _context15.abrupt("return", {
                            v: {
                              payload: payload,
                              response: response
                            }
                          });

                        case 15:
                          if (!(((_triggersArray$i7 = triggersArray[i]) === null || _triggersArray$i7 === void 0 ? void 0 : _triggersArray$i7.shouldReturnResponse) === true)) {
                            _context15.next = 17;
                            break;
                          }

                          return _context15.abrupt("return", {
                            v: response
                          });

                        case 17:
                          return _context15.abrupt("break", 73);

                        case 18:
                          if (_this2.ensureWeb3IsInstalled()) {
                            _context15.next = 20;
                            break;
                          }

                          throw new Error(ERROR_WEB3_MISSING);

                        case 20:
                          if (triggersArray[i].message) {
                            _context15.next = 22;
                            break;
                          }

                          throw new Error('web3Sign trigger does not have a message to sign');

                        case 22:
                          if (!(!triggersArray[i].signer || !_this2.web3.utils.isAddress(triggersArray[i].signer))) {
                            _context15.next = 24;
                            break;
                          }

                          throw new Error('web3Sign trigger signer address missing or invalid');

                        case 24:
                          if (!(((_triggersArray$i8 = triggersArray[i]) === null || _triggersArray$i8 === void 0 ? void 0 : _triggersArray$i8.shouldAwait) === true)) {
                            _context15.next = 28;
                            break;
                          }

                          _context15.next = 27;
                          return _this2.web3.eth.personal.sign(triggersArray[i].message, triggersArray[i].signer);

                        case 27:
                          response = _context15.sent;

                        case 28:
                          // Sign a message using web3 (does NOT await)
                          if (((_triggersArray$i9 = triggersArray[i]) === null || _triggersArray$i9 === void 0 ? void 0 : _triggersArray$i9.shouldAwait) === false) response = _this2.web3.eth.personal.sign(triggersArray[i].message, triggersArray[i].signer); // Save response

                          if (((_triggersArray$i10 = triggersArray[i]) === null || _triggersArray$i10 === void 0 ? void 0 : _triggersArray$i10.saveResponse) === true) _this2.memoryCard.save(response); // Return payload and response

                          if (!(((_triggersArray$i11 = triggersArray[i]) === null || _triggersArray$i11 === void 0 ? void 0 : _triggersArray$i11.shouldReturnPayload) === true)) {
                            _context15.next = 32;
                            break;
                          }

                          return _context15.abrupt("return", {
                            v: {
                              payload: payload,
                              response: response
                            }
                          });

                        case 32:
                          if (!(((_triggersArray$i12 = triggersArray[i]) === null || _triggersArray$i12 === void 0 ? void 0 : _triggersArray$i12.shouldReturnResponse) === true)) {
                            _context15.next = 34;
                            break;
                          }

                          return _context15.abrupt("return", {
                            v: response
                          });

                        case 34:
                          return _context15.abrupt("break", 73);

                        case 35:
                          if (triggersArray[i].pluginName) {
                            _context15.next = 37;
                            break;
                          }

                          throw new Error('callPluginEndpoint trigger does not have an plugin name to call');

                        case 37:
                          if (triggersArray[i].endpoint) {
                            _context15.next = 39;
                            break;
                          }

                          throw new Error('callPluginEndpoint trigger does not have an endpoint to call');

                        case 39:
                          if (!(((_triggersArray$i13 = triggersArray[i]) === null || _triggersArray$i13 === void 0 ? void 0 : _triggersArray$i13.shouldAwait) === true)) {
                            _context15.next = 44;
                            break;
                          } // Check if a saved response has to be used to fill a parameter needed by the plugin


                          if (triggersArray[i].useSavedResponse === true) {
                            triggersArray[i].params[triggersArray[i].savedResponseAs] = _this2.memoryCard.get(triggersArray[i].savedResponseAt);
                          } // Call the endpoint


                          _context15.next = 43;
                          return (0, _Cloud.run)((0, _concat.default)(_context11 = "".concat(triggersArray[i].pluginName, "_")).call(_context11, triggersArray[i].endpoint), triggersArray[i].params);

                        case 43:
                          response = _context15.sent;

                        case 44:
                          // Call a plugin endpoint (does NOT await)
                          if (((_triggersArray$i14 = triggersArray[i]) === null || _triggersArray$i14 === void 0 ? void 0 : _triggersArray$i14.shouldAwait) === false) {
                            // Check if a saved response has to be used to fill a parameter needed by the plugin
                            if (triggersArray[i].useSavedResponse === true) {
                              triggersArray[i].params[triggersArray[i].savedResponseAs] = _this2.memoryCard.get(triggersArray[i].savedResponseAt);
                            } // Call the endpoint


                            response = (0, _Cloud.run)((0, _concat.default)(_context12 = "".concat(triggersArray[i].pluginName, "_")).call(_context12, triggersArray[i].endpoint), triggersArray[i].params);
                          } // If the response contains a trigger, run it


                          if (!(triggersArray[i].runResponseTrigger === true)) {
                            _context15.next = 49;
                            break;
                          }

                          _context15.next = 48;
                          return _this2.handleTriggers(response.data.result.triggers, response.data.result.data);

                        case 48:
                          response = _context15.sent;

                        case 49:
                          // Save response
                          if (((_triggersArray$i15 = triggersArray[i]) === null || _triggersArray$i15 === void 0 ? void 0 : _triggersArray$i15.saveResponse) === true) _this2.memoryCard.save(response); // If should not run the response trigger, continues the loop and does not return (to avoid breaking the loop execution and run other pending triggers)

                          if (!(((_triggersArray$i16 = triggersArray[i]) === null || _triggersArray$i16 === void 0 ? void 0 : _triggersArray$i16.runResponseTrigger) === false)) {
                            _context15.next = 52;
                            break;
                          }

                          return _context15.abrupt("return", "continue");

                        case 52:
                          if (!(((_triggersArray$i17 = triggersArray[i]) === null || _triggersArray$i17 === void 0 ? void 0 : _triggersArray$i17.shouldReturnPayload) === true)) {
                            _context15.next = 54;
                            break;
                          }

                          return _context15.abrupt("return", {
                            v: {
                              payload: 'payload',
                              response: response
                            }
                          });

                        case 54:
                          if (!(((_triggersArray$i18 = triggersArray[i]) === null || _triggersArray$i18 === void 0 ? void 0 : _triggersArray$i18.shouldReturnResponse) === true)) {
                            _context15.next = 56;
                            break;
                          }

                          return _context15.abrupt("return", {
                            v: response
                          });

                        case 56:
                          return _context15.abrupt("break", 73);

                        case 57:
                          if (_this2.ensureWeb3IsInstalled()) {
                            _context15.next = 59;
                            break;
                          }

                          throw new Error(ERROR_WEB3_MISSING);

                        case 59:
                          if (triggersArray[i].parameters) {
                            _context15.next = 61;
                            break;
                          }

                          throw new Error('web3SignV4 trigger does not have `parameters` to sign');

                        case 61:
                          if (triggersArray[i].from) {
                            _context15.next = 63;
                            break;
                          }

                          throw new Error('web3SignV4 trigger does not have a `from` address');

                        case 63:
                          if (!(((_triggersArray$i19 = triggersArray[i]) === null || _triggersArray$i19 === void 0 ? void 0 : _triggersArray$i19.shouldAwait) === true)) {
                            _context15.next = 66;
                            break;
                          }

                          _context15.next = 66;
                          return _this2.web3.currentProvider.send({
                            method: 'eth_signTypedData_v4',
                            params: triggersArray[i].parameters,
                            from: triggersArray[i].from
                          },
                          /*#__PURE__*/
                          // eslint-disable-next-line no-loop-func
                          function () {
                            var _ref2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee9(error, result) {
                              var _triggersArray$i20;

                              return _regenerator.default.wrap(function (_context13) {
                                while (1) {
                                  switch (_context13.prev = _context13.next) {
                                    case 0:
                                      if (!error) {
                                        _context13.next = 2;
                                        break;
                                      }

                                      throw new Error(error.message || error);

                                    case 2:
                                      // Save response
                                      if (((_triggersArray$i20 = triggersArray[i]) === null || _triggersArray$i20 === void 0 ? void 0 : _triggersArray$i20.saveResponse) === true) _this2.memoryCard.save(result);
                                      response = result;

                                    case 4:
                                    case "end":
                                      return _context13.stop();
                                  }
                                }
                              }, _callee9);
                            }));

                            return function () {
                              return _ref2.apply(this, arguments);
                            };
                          }());

                        case 66:
                          if (((_triggersArray$i21 = triggersArray[i]) === null || _triggersArray$i21 === void 0 ? void 0 : _triggersArray$i21.shouldAwait) === false) {
                            _this2.web3.currentProvider.send({
                              method: 'eth_signTypedData_v4',
                              params: triggersArray[i].parameters,
                              from: triggersArray[i].from
                            },
                            /*#__PURE__*/
                            // eslint-disable-next-line no-loop-func
                            function () {
                              var _ref3 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee10(error, result) {
                                var _triggersArray$i22;

                                return _regenerator.default.wrap(function (_context14) {
                                  while (1) {
                                    switch (_context14.prev = _context14.next) {
                                      case 0:
                                        if (!error) {
                                          _context14.next = 2;
                                          break;
                                        }

                                        throw new Error(error.message || error);

                                      case 2:
                                        // Save response
                                        if (((_triggersArray$i22 = triggersArray[i]) === null || _triggersArray$i22 === void 0 ? void 0 : _triggersArray$i22.saveResponse) === true) _this2.memoryCard.save(result);
                                        response = result;

                                      case 4:
                                      case "end":
                                        return _context14.stop();
                                    }
                                  }
                                }, _callee10);
                              }));

                              return function () {
                                return _ref3.apply(this, arguments);
                              };
                            }());
                          } // Return payload and response


                          if (!(((_triggersArray$i23 = triggersArray[i]) === null || _triggersArray$i23 === void 0 ? void 0 : _triggersArray$i23.shouldReturnPayload) === true)) {
                            _context15.next = 69;
                            break;
                          }

                          return _context15.abrupt("return", {
                            v: {
                              payload: payload,
                              response: response
                            }
                          });

                        case 69:
                          if (!(((_triggersArray$i24 = triggersArray[i]) === null || _triggersArray$i24 === void 0 ? void 0 : _triggersArray$i24.shouldReturnResponse) === true)) {
                            _context15.next = 71;
                            break;
                          }

                          return _context15.abrupt("return", {
                            v: response
                          });

                        case 71:
                          return _context15.abrupt("break", 73);

                        case 72:
                          throw new Error("Unknown trigger: \"".concat((_triggersArray$i25 = triggersArray[i]) === null || _triggersArray$i25 === void 0 ? void 0 : _triggersArray$i25.name, "\""));

                        case 73:
                        case "end":
                          return _context15.stop();
                      }
                    }
                  }, _loop);
                });
                i = 0;

              case 4:
                if (!(i < triggersArray.length)) {
                  _context16.next = 14;
                  break;
                }

                return _context16.delegateYield(_loop(i), "t0", 6);

              case 6:
                _ret = _context16.t0;

                if (!(_ret === "continue")) {
                  _context16.next = 9;
                  break;
                }

                return _context16.abrupt("continue", 11);

              case 9:
                if (!((0, _typeof2.default)(_ret) === "object")) {
                  _context16.next = 11;
                  break;
                }

                return _context16.abrupt("return", _ret.v);

              case 11:
                i++;
                _context16.next = 4;
                break;

              case 14:
                // Delete all saved data
                this.memoryCard.deleteSaved();

              case 15:
              case "end":
                return _context16.stop();
            }
          }
        }, _callee11, this);
      }));

      return function () {
        return _handleTriggers.apply(this, arguments);
      };
    }()
  }, {
    key: "getAllERC20",
    value: function () {
      var _getAllERC = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee12() {
        var _ref4,
            chain,
            address,
            result,
            _args13 = arguments;

        return _regenerator.default.wrap(function (_context17) {
          while (1) {
            switch (_context17.prev = _context17.next) {
              case 0:
                _ref4 = _args13.length > 0 && _args13[0] !== undefined ? _args13[0] : {}, chain = _ref4.chain, address = _ref4.address;
                _context17.next = 3;
                return (0, _Cloud.run)('getAllERC20', {
                  chain: chain,
                  address: address
                });

              case 3:
                result = _context17.sent;
                return _context17.abrupt("return", result);

              case 5:
              case "end":
                return _context17.stop();
            }
          }
        }, _callee12);
      }));

      return function () {
        return _getAllERC.apply(this, arguments);
      };
    }()
  }, {
    key: "getERC20",
    value: function () {
      var _getERC = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee13() {
        var _ref5,
            chain,
            address,
            symbol,
            tokenAddress,
            result,
            _args14 = arguments;

        return _regenerator.default.wrap(function (_context18) {
          while (1) {
            switch (_context18.prev = _context18.next) {
              case 0:
                _ref5 = _args14.length > 0 && _args14[0] !== undefined ? _args14[0] : {}, chain = _ref5.chain, address = _ref5.address, symbol = _ref5.symbol, tokenAddress = _ref5.tokenAddress;
                result = (0, _Cloud.run)('getERC20', {
                  chain: chain,
                  address: address,
                  symbol: symbol,
                  tokenAddress: tokenAddress
                });
                return _context18.abrupt("return", result);

              case 3:
              case "end":
                return _context18.stop();
            }
          }
        }, _callee13);
      }));

      return function () {
        return _getERC.apply(this, arguments);
      };
    }()
  }, {
    key: "getNFTs",
    value: function () {
      var _ref6 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref6$chain = _ref6.chain,
          chain = _ref6$chain === void 0 ? 'Eth' : _ref6$chain,
          _ref6$address = _ref6.address,
          address = _ref6$address === void 0 ? '' : _ref6$address;

      return (0, _Cloud.run)('getNFTs_old', {
        chain: chain,
        address: address
      });
    }
  }, {
    key: "getNFTsCount",
    value: function () {
      var _ref7 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref7$chain = _ref7.chain,
          chain = _ref7$chain === void 0 ? 'Eth' : _ref7$chain,
          _ref7$address = _ref7.address,
          address = _ref7$address === void 0 ? '' : _ref7$address;

      return (0, _Cloud.run)('getNFTsCount_old', {
        chain: chain,
        address: address
      });
    }
  }, {
    key: "getTransactions",
    value: function () {
      var _ref8 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref8$chain = _ref8.chain,
          chain = _ref8$chain === void 0 ? 'Eth' : _ref8$chain,
          _ref8$address = _ref8.address,
          address = _ref8$address === void 0 ? '' : _ref8$address,
          _ref8$order = _ref8.order,
          order = _ref8$order === void 0 ? 'desc' : _ref8$order;

      return (0, _Cloud.run)('getTransactions', {
        chain: chain,
        address: address,
        order: order
      });
    }
  }, {
    key: "getTransactionsCount",
    value: function () {
      var _ref9 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref9$chain = _ref9.chain,
          chain = _ref9$chain === void 0 ? 'Eth' : _ref9$chain,
          _ref9$address = _ref9.address,
          address = _ref9$address === void 0 ? '' : _ref9$address;

      return (0, _Cloud.run)('getTransactionsCount', {
        chain: chain,
        address: address
      });
    }
  }, {
    key: "transfer",
    value: function () {
      var _transfer = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee14() {
        var _ref10,
            _ref10$type,
            type,
            _ref10$receiver,
            receiver,
            _ref10$contractAddres,
            contractAddress,
            contract_address,
            _ref10$amount,
            amount,
            _ref10$tokenId,
            tokenId,
            token_id,
            _ref10$system,
            system,
            _ref10$awaitReceipt,
            awaitReceipt,
            options,
            web3,
            sender,
            transferOperation,
            customToken,
            transferEvents,
            _args15 = arguments;

        return _regenerator.default.wrap(function (_context19) {
          while (1) {
            switch (_context19.prev = _context19.next) {
              case 0:
                _ref10 = _args15.length > 0 && _args15[0] !== undefined ? _args15[0] : {}, _ref10$type = _ref10.type, type = _ref10$type === void 0 ? 'native' : _ref10$type, _ref10$receiver = _ref10.receiver, receiver = _ref10$receiver === void 0 ? '' : _ref10$receiver, _ref10$contractAddres = _ref10.contractAddress, contractAddress = _ref10$contractAddres === void 0 ? '' : _ref10$contractAddres, contract_address = _ref10.contract_address, _ref10$amount = _ref10.amount, amount = _ref10$amount === void 0 ? '' : _ref10$amount, _ref10$tokenId = _ref10.tokenId, tokenId = _ref10$tokenId === void 0 ? '' : _ref10$tokenId, token_id = _ref10.token_id, _ref10$system = _ref10.system, system = _ref10$system === void 0 ? 'evm' : _ref10$system, _ref10$awaitReceipt = _ref10.awaitReceipt, awaitReceipt = _ref10$awaitReceipt === void 0 ? true : _ref10$awaitReceipt; // Allow snake-case for backwards compatibility
                // eslint-disable-next-line camelcase

                contractAddress = contractAddress || contract_address; // eslint-disable-next-line camelcase

                tokenId = tokenId || token_id;
                options = {
                  receiver: receiver,
                  contractAddress: contractAddress,
                  amount: amount,
                  tokenId: tokenId,
                  system: system,
                  awaitReceipt: awaitReceipt
                };

                _TransferUtils.default.isSupportedType(type);

                _TransferUtils.default.validateInput(type, options);

                if (this.ensureWeb3IsInstalled()) {
                  _context19.next = 8;
                  break;
                }

                throw new Error(ERROR_WEB3_MISSING);

              case 8:
                web3 = this.web3;
                _context19.next = 11;
                return web3.eth.getAccounts();

              case 11:
                _context19.next = 13;
                return _context19.sent[0];

              case 13:
                sender = _context19.sent;

                if (sender) {
                  _context19.next = 16;
                  break;
                }

                throw new Error('Sender address not found');

              case 16:
                if (tokenId) _TransferUtils.default.isUint256(tokenId);
                if (type !== 'native') customToken = new web3.eth.Contract(_TransferUtils.default.abi[type], contractAddress);
                _context19.t0 = type;
                _context19.next = _context19.t0 === 'native' ? 21 : _context19.t0 === 'erc20' ? 23 : _context19.t0 === 'erc721' ? 25 : _context19.t0 === 'erc1155' ? 27 : 29;
                break;

              case 21:
                transferOperation = web3.eth.sendTransaction({
                  from: sender,
                  to: receiver,
                  value: amount
                });
                return _context19.abrupt("break", 30);

              case 23:
                transferOperation = customToken.methods.transfer(receiver, amount).send({
                  from: sender
                });
                return _context19.abrupt("break", 30);

              case 25:
                transferOperation = customToken.methods.safeTransferFrom(sender, receiver, "".concat(tokenId)).send({
                  from: sender
                });
                return _context19.abrupt("break", 30);

              case 27:
                transferOperation = customToken.methods.safeTransferFrom(sender, receiver, "".concat(tokenId), amount, '0x').send({
                  from: sender
                });
                return _context19.abrupt("break", 30);

              case 29:
                throw new Error("Unknown transfer type: \"".concat(type, "\""));

              case 30:
                if (!awaitReceipt) {
                  _context19.next = 32;
                  break;
                }

                return _context19.abrupt("return", transferOperation);

              case 32:
                transferEvents = new EventEmitter();
                transferOperation.on('transactionHash', function (hash) {
                  transferEvents.emit('transactionHash', hash);
                }).on('receipt', function (receipt) {
                  transferEvents.emit('receipt', receipt);
                }).on('confirmation', function (confirmationNumber, receipt) {
                  transferEvents.emit('confirmation', (confirmationNumber, receipt));
                }).on('error', function (error) {
                  transferEvents.emit('error', error);
                  throw error;
                });
                return _context19.abrupt("return", transferEvents);

              case 35:
              case "end":
                return _context19.stop();
            }
          }
        }, _callee14, this);
      }));

      return function () {
        return _transfer.apply(this, arguments);
      };
    }()
  }, {
    key: "executeFunction",
    value: function () {
      var _executeFunction = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee15() {
        var _context20;

        var _ref11,
            contractAddress,
            abi,
            functionName,
            msgValue,
            _ref11$awaitReceipt,
            awaitReceipt,
            _ref11$params,
            params,
            web3,
            contractOptions,
            functionData,
            stateMutability,
            isReadFunction,
            currentAddress,
            errors,
            _iterator,
            _step,
            input,
            value,
            parsedInputs,
            contract,
            customFunction,
            response,
            contractExecuteEvents,
            _args16 = arguments;

        return _regenerator.default.wrap(function (_context21) {
          while (1) {
            switch (_context21.prev = _context21.next) {
              case 0:
                _ref11 = _args16.length > 0 && _args16[0] !== undefined ? _args16[0] : {}, contractAddress = _ref11.contractAddress, abi = _ref11.abi, functionName = _ref11.functionName, msgValue = _ref11.msgValue, _ref11$awaitReceipt = _ref11.awaitReceipt, awaitReceipt = _ref11$awaitReceipt === void 0 ? true : _ref11$awaitReceipt, _ref11$params = _ref11.params, params = _ref11$params === void 0 ? {} : _ref11$params;

                if (this.ensureWeb3IsInstalled()) {
                  _context21.next = 3;
                  break;
                }

                throw new Error(ERROR_WEB3_MISSING);

              case 3:
                web3 = this.web3;
                contractOptions = {};
                functionData = (0, _find.default)(abi).call(abi, function (x) {
                  return x.name === functionName;
                });

                if (functionData) {
                  _context21.next = 8;
                  break;
                }

                throw new Error('Function does not exist in abi');

              case 8:
                stateMutability = functionData === null || functionData === void 0 ? void 0 : functionData.stateMutability;
                isReadFunction = stateMutability === 'view' || stateMutability === 'pure';

                if (isReadFunction) {
                  _context21.next = 20;
                  break;
                }

                if (params.from) {
                  _context21.next = 20;
                  break;
                }

                _context21.next = 14;
                return web3.eth.getAccounts();

              case 14:
                _context21.next = 16;
                return _context21.sent[0];

              case 16:
                currentAddress = _context21.sent;

                if (currentAddress) {
                  _context21.next = 19;
                  break;
                }

                throw new Error('From address is required');

              case 19:
                contractOptions.from = currentAddress;

              case 20:
                errors = [];
                _iterator = _createForOfIteratorHelper(functionData.inputs);

                try {
                  for (_iterator.s(); !(_step = _iterator.n()).done;) {
                    input = _step.value;
                    value = params[input.name];

                    if (!value && typeof value !== 'number' && typeof value !== 'boolean') {
                      errors.push("".concat(input.name, " is required"));
                    }
                  }
                } catch (err) {
                  _iterator.e(err);
                } finally {
                  _iterator.f();
                }

                if (!(errors.length > 0)) {
                  _context21.next = 25;
                  break;
                }

                throw errors;

              case 25:
                parsedInputs = (0, _map.default)(_context20 = functionData.inputs).call(_context20, function (x) {
                  return params[x.name];
                });
                contract = new web3.eth.Contract(abi, contractAddress, contractOptions);
                customFunction = contract.methods[functionName];
                response = isReadFunction ? customFunction.apply(void 0, (0, _toConsumableArray2.default)((0, _values.default)(parsedInputs))).call() : customFunction.apply(void 0, (0, _toConsumableArray2.default)((0, _values.default)(parsedInputs))).send(msgValue ? {
                  value: msgValue
                } : null);

                if (!awaitReceipt) {
                  _context21.next = 31;
                  break;
                }

                return _context21.abrupt("return", response);

              case 31:
                contractExecuteEvents = new EventEmitter();
                response.on('transactionHash', function (hash) {
                  contractExecuteEvents.emit('transactionHash', hash);
                }).on('receipt', function (receipt) {
                  contractExecuteEvents.emit('receipt', receipt);
                }).on('confirmation', function (confirmationNumber, receipt) {
                  contractExecuteEvents.emit('confirmation', (confirmationNumber, receipt));
                }).on('error', function (error) {
                  contractExecuteEvents.emit('error', error);
                  throw error;
                });
                return _context21.abrupt("return", contractExecuteEvents);

              case 34:
              case "end":
                return _context21.stop();
            }
          }
        }, _callee15, this);
      }));

      return function () {
        return _executeFunction.apply(this, arguments);
      };
    }()
  }, {
    key: "getSigningData",
    value: function () {
      return "Moralis Authentication"; // const data = `Moralis Authentication`;
      // return data;
    }
  }, {
    key: "on",
    value: function (eventName, cb) {
      var _window = window,
          ethereum = _window.ethereum;

      if (!ethereum || !ethereum.on) {
        // eslint-disable-next-line no-console
        console.warn(WARNING);
        return function () {
          // eslint-disable-next-line no-console
          console.warn(WARNING);
        };
      }

      ethereum.on(eventName, cb);
      return function () {
        // eslint-disable-next-line no-console
        console.warn('UNSUB NOT SUPPORTED');
      };
    }
  }, {
    key: "getChainId",
    value: function () {
      var _getChainId = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee16() {
        return _regenerator.default.wrap(function (_context22) {
          while (1) {
            switch (_context22.prev = _context22.next) {
              case 0:
                if (!this.ensureWeb3IsInstalled()) {
                  _context22.next = 4;
                  break;
                }

                _context22.next = 3;
                return this.web3.eth.net.getId();

              case 3:
                return _context22.abrupt("return", _context22.sent);

              case 4:
                throw new Error(ERROR_WEB3_MISSING);

              case 5:
              case "end":
                return _context22.stop();
            }
          }
        }, _callee16, this);
      }));

      return function () {
        return _getChainId.apply(this, arguments);
      };
    }()
  }, {
    key: "ensureWeb3IsInstalled",
    value: function () {
      return this.web3 ? true : false;
    }
  }, {
    key: "isMetaMaskInstalled",
    value: function () {
      var _isMetaMaskInstalled = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee17() {
        return _regenerator.default.wrap(function (_context23) {
          while (1) {
            switch (_context23.prev = _context23.next) {
              case 0:
                _context23.next = 2;
                return (0, _detectProvider.default)();

              case 2:
                if (!_context23.sent) {
                  _context23.next = 6;
                  break;
                }

                _context23.t0 = true;
                _context23.next = 7;
                break;

              case 6:
                _context23.t0 = false;

              case 7:
                return _context23.abrupt("return", _context23.t0);

              case 8:
              case "end":
                return _context23.stop();
            }
          }
        }, _callee17);
      }));

      return function () {
        return _isMetaMaskInstalled.apply(this, arguments);
      };
    }()
  }, {
    key: "switchNetwork",
    value: function () {
      var _switchNetwork = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee18(chainId) {
        var currentNetwork;
        return _regenerator.default.wrap(function (_context24) {
          while (1) {
            switch (_context24.prev = _context24.next) {
              case 0:
                chainId = verifyChainId(chainId); // Check if the user wallet is already on `chainId`

                _context24.t0 = fromDecimalToHex;
                _context24.next = 4;
                return this.getChainId();

              case 4:
                _context24.t1 = _context24.sent;
                currentNetwork = (0, _context24.t0)(_context24.t1);

                if (!(currentNetwork === chainId)) {
                  _context24.next = 8;
                  break;
                }

                return _context24.abrupt("return");

              case 8:
                _context24.next = 10;
                return window.ethereum.request({
                  method: 'wallet_switchEthereumChain',
                  params: [{
                    chainId: chainId
                  }]
                });

              case 10:
              case "end":
                return _context24.stop();
            }
          }
        }, _callee18, this);
      }));

      return function () {
        return _switchNetwork.apply(this, arguments);
      };
    }()
  }, {
    key: "addNetwork",
    value: function () {
      var _addNetwork = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee19(chainId, chainName, currencyName, currencySymbol, rpcUrl, blockExplorerUrl) {
        return _regenerator.default.wrap(function (_context25) {
          while (1) {
            switch (_context25.prev = _context25.next) {
              case 0:
                chainId = verifyChainId(chainId);
                _context25.next = 3;
                return window.ethereum.request({
                  method: 'wallet_addEthereumChain',
                  params: [{
                    chainId: chainId,
                    chainName: chainName,
                    nativeCurrency: {
                      name: currencyName,
                      symbol: currencySymbol,
                      decimals: 18
                    },
                    rpcUrls: [rpcUrl],
                    blockExplorerUrls: [blockExplorerUrl]
                  }]
                });

              case 3:
              case "end":
                return _context25.stop();
            }
          }
        }, _callee19);
      }));

      return function () {
        return _addNetwork.apply(this, arguments);
      };
    }()
  }]);
  return MoralisWeb3;
}();

(0, _defineProperty2.default)(MoralisWeb3, "speedyNodeApiKey", void 0);
(0, _defineProperty2.default)(MoralisWeb3, "memoryCard", {
  save: function (what) {
    this.saved = what;
  },
  get: function (where) {
    if (!this.saved) throw new Error('Nothing saved to memory card'); // In case the saved data is not an object but a simple string or number

    if (where.length === 0) return this.getSaved();
    var tmp;
    var savedTmp = this.saved;

    for (var i = 0; i < where.length; i++) {
      tmp = savedTmp[where[i]];
      savedTmp = tmp;
    }

    return savedTmp;
  },
  getSaved: function () {
    return this.saved;
  },
  deleteSaved: function () {
    this.saved = undefined;
  }
});

function fromDecimalToHex(number) {
  if (typeof number !== 'number') throw 'The input provided should be a number';
  return "0x".concat(number.toString(16));
}

function verifyChainId(chainId) {
  // Check if chainId is a number, in that case convert to hex
  if (typeof chainId === 'number') chainId = fromDecimalToHex(chainId);
  return chainId;
}

MoralisWeb3.onConnect = (0, _bind.default)(_context26 = MoralisWeb3.on).call(_context26, MoralisWeb3, EthereumEvents.CONNECT);
MoralisWeb3.onDisconnect = (0, _bind.default)(_context27 = MoralisWeb3.on).call(_context27, MoralisWeb3, EthereumEvents.DISCONNECT);
MoralisWeb3.onChainChanged = (0, _bind.default)(_context28 = MoralisWeb3.on).call(_context28, MoralisWeb3, EthereumEvents.CHAIN_CHANGED);
MoralisWeb3.onAccountsChanged = (0, _bind.default)(_context29 = MoralisWeb3.on).call(_context29, MoralisWeb3, EthereumEvents.ACCOUNTS_CHANGED);
var _default = MoralisWeb3;
exports.default = _default;
